kratos/utilities/function_parser_utility.cpp

```diff
@@ -0,0 +1,184 @@
+        mFunctionBody.find(std::string("y")) == std::string::npos &&
+        mFunctionBody.find(std::string("z")) == std::string::npos) {
+        mDependsOnSpace = false;
+    }
+}
+
+/***********************************************************************************/
+/***********************************************************************************/
+
+GenericFunctionUtility::GenericFunctionUtility(GenericFunctionUtility const& rOther)
+    : mNameSpace(rOther.mNameSpace),
+      mFunctionBody(rOther.mFunctionBody),
+      mDependsOnSpace(rOther.mDependsOnSpace),
+      mUseLocalSystem(rOther.mUseLocalSystem),
+      mRotationMatrix(rOther.mRotationMatrix),
+      mCenterCoordinates(rOther.mCenterCoordinates)
+{
+    // Initialize parser classes
+    InitializeParser();
+}
+
+/***********************************************************************************/
+/***********************************************************************************/
+
+GenericFunctionUtility::~GenericFunctionUtility()
+{
+    te_free(mpTinyExpr);
+}
+
+/***********************************************************************************/
+/***********************************************************************************/
+
+std::string GenericFunctionUtility::FunctionBody()
+{
+    std::string aux_string = mFunctionBody;
+    std::unordered_map<std::string,std::string> aux_replace_letters({{"**","^"},{"X","x0"},{"Y","y0"},{"Z","z0"}});
+    for (auto& r_pair : aux_replace_letters) {
+         aux_string = StringUtilities::ReplaceAllSubstrings(aux_string, r_pair.second, r_pair.first);
+    }
+    return aux_string;
+}
+
+/***********************************************************************************/
+/***********************************************************************************/
+
+double GenericFunctionUtility::RotateAndCallFunction(
+    const double x,
+    const double y,
+    const double z,
+    const double t,
+    const double X,
+    const double Y,
+    const double Z
+    )
+{
+    array_1d<double,3> xglobal;
+    xglobal[0] = x;
+    xglobal[1] = y;
+    xglobal[2] = z;
+    array_1d<double,3> xlocal = prod(mRotationMatrix, (xglobal - mCenterCoordinates) );
+    array_1d<double,3> xglobal_initial;
+    xglobal_initial[0] = X;
+    xglobal_initial[1] = Y;
+    xglobal_initial[2] = Z;
+    array_1d<double,3> xlocal_initial = prod(mRotationMatrix, (xglobal_initial - mCenterCoordinates) );
+    return CallFunction(xlocal[0],xlocal[1],xlocal[2], t,xlocal_initial[0],xlocal_initial[1],xlocal_initial[2]);
+}
+
+/***********************************************************************************/
+/***********************************************************************************/
+
+double GenericFunctionUtility::CallFunction(
+    const double x,
+    const double y,
+    const double z,
+    const double t,
+    const double X,
+    const double Y,
+    const double Z
+    )
+{
+    mNameSpace["x"] = x;
```

Comment:
@RiccardoRossi it is used here, the alternative is to build the function each call!

Reply:
but the point is that the mNameSpace is not needed at all. It is an intermediate data structure that i think can be avoided completely